### PR TITLE
Update Docs to Show Attributes

### LINKED ISSUE #1653: Docs failing to reflect attributes
Certain classes such as `ObjectiveBase` and `ComponentBase` aren't reflecting their attributes. `ObjectiveBase` has `@property` and `@classproperty` decorators but none are displayed, while `ComponentBase` additionally has a class attribute with nothing being displayed.
https://evalml.alteryx.com/en/stable/generated/evalml.objectives.ObjectiveBase.html
https://evalml.alteryx.com/en/stable/generated/evalml.pipelines.components.ComponentBase.html?highlight=Componentbase

However, `PipelineBase` has instance attributes and class attributes reflected in its docs.
https://evalml.alteryx.com/en/stable/generated/evalml.pipelines.PipelineBase.html#evalml.pipelines.PipelineBase

### PR DESCRIPTION
**Pull request by ParthivNaresh**
Fixes #1653

### CODE REVIEW

File: docs/source/_templates/data_check_class.rst [hunk 1]
```diff
@@ -12,19 +12,6 @@
    .. autoattribute:: name
 
    {% endblock %}
-
```
Comment (ParthivNaresh) [inline docs/source/_templates/data_check_class.rst:15]: DataCheck subclasses don't have any instance attributes so this section is unnecessary.

File: docs/source/_templates/objective_base_class.rst [hunk 2]
```diff
@@ -0,0 +1,31 @@
+{{ fullname | escape | underline}}
```
Comment (ParthivNaresh) [inline docs/source/_templates/objective_base_class.rst:1]: Previously `class.rst` was the default template for Objective subclasses and it only included methods and were missing class attributes. This is a new `rst` for the subclass.

File: docs/source/_templates/objective_class.rst [hunk 3]
```diff
@@ -0,0 +1,37 @@
+{{ fullname | escape | underline}}
```
Comment (ParthivNaresh) [inline docs/source/_templates/objective_class.rst:1]: Previously `class.rst` was the default template for `Objective` subclasses and it only included methods and were missing class attributes. This is a new rst for the subclass.

File: docs/source/api_reference.rst [hunk 4]
```diff
@@ -36,6 +36,30 @@ Utilities to preprocess data before using evalml.
     split_data
 
 
+.. currentmodule:: evalml.exceptions
```
Comment (ParthivNaresh) [inline docs/source/api_reference.rst:39]: Adding Exceptions to the docs

File: docs/source/conf.py [hunk 5]
```diff
@@ -207,6 +207,7 @@
     'user': ('https://github.com/%s', '@')
 }
 
+#autoclass_content = 'both'
```
Comment (ParthivNaresh) [inline docs/source/conf.py:210]: Concatenates both the class and init method's docstrings. Wanted to know if anyone had a particular opinion on whether this should be included or not!

File: evalml/objectives/objective_base.py [hunk 6]
```diff
@@ -10,6 +10,8 @@
 class ObjectiveBase(ABC):
     """Base class for all objectives."""
 
+    problem_types = None
```
Comment (ParthivNaresh) [inline evalml/objectives/objective_base.py:13]: To allow ObjectiveBase to fall under the same `objective_base_class.rst` template as the other base classes

File: evalml/objectives/standard_metrics.py [hunk 7]
```diff
@@ -365,6 +365,7 @@ class MAPE(TimeSeriesRegressionObjective):
     """
     name = "Mean Absolute Percentage Error"
     greater_is_better = False
+    score_needs_proba = False
```
Comment (ParthivNaresh) [inline evalml/objectives/standard_metrics.py:368]: To allow `MAPE` to fall under the same `objective_class.rst` as the other metrics

File: evalml/pipelines/components/estimators/estimator.py [hunk 8]
```diff
@@ -23,6 +24,7 @@ class Estimator(ComponentBase):
     """
     # We can't use the inspect module to dynamically determine this because of issue 1582
     predict_uses_y = False
+    model_family = ModelFamily.NONE
```
Comment (ParthivNaresh) [inline evalml/pipelines/components/estimators/estimator.py:27]: Wanted some feedback on if this is ok? It subclasses the `model_family` abstract method and is a class property. If it's fine to add this, I'll probably add a custom `component.rst` template

File: docs/source/api_reference.rst [hunk 9]
```diff
@@ -36,6 +36,30 @@ Utilities to preprocess data before using evalml.
...
+.. autosummary::
+    :toctree: generated
+    :template: class.rst
+    :nosignatures:
+
+    MethodPropertyNotFoundError
+    PipelineNotFoundError
+    ObjectiveNotFoundError
+    IllFormattedClassNameError
+    MissingComponentError
```
Comment (bchen1116) [inline docs/source/api_reference.rst:53]: Nice! nit-pick: Is it possible to fix some of the formatting of the docstrings in these exceptions? Ie `all_components()` here:
image

File: docs/source/conf.py (hunk 5 above)
Comment (bchen1116) [reply docs/source/conf.py:210]: Do you have examples on what this looks like in the docs?

File: evalml/objectives/binary_classification_objective.py [hunk 10]
```diff
@@ -7,11 +7,7 @@
 
 
 class BinaryClassificationObjective(ObjectiveBase):
-    """Base class for all binary classification objectives.
-
-    problem_types (list(ProblemType)): List of problem types that this objective is defined for. Set to [ProblemTypes.BINARY].
```
Comment (bchen1116) [inline evalml/objectives/binary_classification_objective.py:12]: Why did this get removed?

File: evalml/pipelines/components/estimators/estimator.py (hunk 8 above)
Comment (bchen1116) [reply evalml/pipelines/components/estimators/estimator.py:27]: Seems like if it doesn't break any tests, it's fine?

**Review by bchen1116 — APPROVED** [review]
This is what the objectiveBase [docs](https://feature-labs-inc-evalml--1673.com.readthedocs.build/en/1673/generated/evalml.objectives.ObjectiveBase.html#evalml.objectives.ObjectiveBase) look like, and componentBase is [here](https://feature-labs-inc-evalml--1673.com.readthedocs.build/en/1673/generated/evalml.pipelines.components.ComponentBase.html#evalml.pipelines.components.ComponentBase)

Docs definitely look better! Left a few comments on improvements, but otherwise looks good to me!

File: evalml/objectives/binary_classification_objective.py (hunk 10 above)
Comment (ParthivNaresh) [reply evalml/objectives/binary_classification_objective.py:12]: @bchen1116 `problem_types` would have been included twice, once as part of the docstring and once as part of the class attributes
Before: https://evalml.alteryx.com/en/stable/generated/evalml.objectives.BinaryClassificationObjective.html?highlight=BinaryClassificationObjective
After: https://evalml.alteryx.com/en/1653-update-docs-to-show-attributes/generated/evalml.objectives.BinaryClassificationObjective.html?highlight=BinaryClassificationObjective

Comment (freddyaboulton) [reply evalml/objectives/binary_classification_objective.py:12]: Maybe we can keep `Base class for all binary classification objectives.`? Agree that the arguments here look somewhat out of place.

File: docs/source/_templates/objective_class.rst [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+.. currentmodule:: {{ module }}
+
+.. autoclass:: {{ objname }}
+   {% set class_attributes = ['name', 'greater_is_better', 'perfect_score', 'positive_only', 'problem_types',
```
Comment (freddyaboulton) [inline docs/source/_templates/objective_class.rst:6]: How come we need to set class_attributes here if we're going to manually add them one-by-one below?

File: evalml/objectives/binary_classification_objective.py (hunk 10 above)
Comment (bchen1116) [reply evalml/objectives/binary_classification_objective.py:12]: Gotcha, thanks

**Review by freddyaboulton — APPROVED** [review]
@ParthivNaresh Looks great to me!

File: docs/source/_templates/objective_class.rst (hunk 11 above)
Comment (ParthivNaresh) [reply docs/source/_templates/objective_class.rst:6]: You're right, it's absolutely unnecessary, let me remove it!

File: docs/source/_templates/data_check_class.rst (hunk 1 above)
Comment (chukarsten) [reply docs/source/_templates/data_check_class.rst:15]: Love it.  Leaving things cleaner than when you found them is an excellent habit.

**Review by chukarsten — APPROVED** [review]

**Pull request merged**